File: ios/AtarasyCore/Sources/AtarasyCore/MemberDigitalDraft.swift
import Foundation

/// Unsent choices only. This value grants no authority and performs no network or key operation.
public struct MemberDigitalDraft: Sendable {
    public enum Choice: String, CaseIterable, Sendable { case undecided, keep, decline }
    public struct Summary: Equatable, Sendable {
        public let goods: Int64
        public let carriage: Int64
        public let total: Int64
    }
    public let approval: MemberApproval
    private var choices: [Data: Choice] = [:]

    public init(approval: MemberApproval) { self.approval = approval }
    public func choice(for id: String) -> Choice { choices[Data(id.utf8)] ?? .undecided }
    public mutating func choose(_ choice: Choice, candidate id: String) throws {
        guard approval.candidates.contains(where: { Data($0.id.utf8) == Data(id.utf8) && $0.valence == "offered" }) else { throw ContractError.invalidTransition }
        choices[Data(id.utf8)] = choice
    }
    public mutating func discard() { choices.removeAll() }

    /// Unknown carriage, incomplete choices, expired terms and unsafe arithmetic never become a zero total.
    public func summary(now: Int64) throws -> Summary {
        guard now >= 0, now < approval.expiresAt,
              approval.mandate.lapsesAt.map({ now < $0 }) ?? true,
              !approval.candidates.isEmpty,
              Set(approval.candidates.map { Data($0.id.utf8) }).count == approval.candidates.count,
              let carriage = approval.carriage, (0...Canonical.maximumInteger).contains(carriage) else { throw ContractError.invalidTransition }
        var goods: Int64 = 0
        for candidate in approval.candidates {
            guard candidate.valence == "offered", choice(for: candidate.id) != .undecided,
                  candidate.quantity > 0, candidate.quantity <= Canonical.maximumInteger,
                  (0...Canonical.maximumInteger).contains(candidate.unitPrice) else { throw ContractError.invalidTransition }
            if choice(for: candidate.id) == .keep && candidate.givenBy == nil {
                let amount = candidate.quantity.multipliedReportingOverflow(by: candidate.unitPrice)
                guard !amount.overflow, amount.partialValue <= Canonical.maximumInteger else { throw ContractError.invalidValue }
                let sum = goods.addingReportingOverflow(amount.partialValue)
                guard !sum.overflow, sum.partialValue <= Canonical.maximumInteger else { throw ContractError.invalidValue }
                goods = sum.partialValue
            }
        }
        let total = goods.addingReportingOverflow(carriage)
        guard !total.overflow, total.partialValue <= Canonical.maximumInteger else { throw ContractError.invalidValue }
        return Summary(goods: goods, carriage: carriage, total: total.partialValue)
    }
}

File: android/app/src/main/java/dev/atarasy/prototype/MemberDigitalDraft.kt
package dev.atarasy.prototype

enum class MemberDigitalChoice { UNDECIDED, KEEP, DECLINE }
data class MemberDigitalSummary(val goods: Long, val carriage: Long, val total: Long)

class MemberDigitalDraft(val approval: MemberApproval) {
    private val choices = mutableMapOf<String, MemberDigitalChoice>()

    fun choice(candidate: String) = choices[candidate] ?: MemberDigitalChoice.UNDECIDED

    fun choose(candidate: String, choice: MemberDigitalChoice) {
        if (approval.candidates.none { it.id == candidate && it.valence == "offered" }) throw MemberFailure.Malformed
        choices[candidate] = choice
    }

    fun discard() = choices.clear()

    fun decisions(now: Long): List<Decision> {
        summary(now)
        return approval.candidates.map { candidate ->
            if (choice(candidate.id) == MemberDigitalChoice.KEEP) Decision(candidate.id, "kept", keptAs = "self")
            else Decision(candidate.id, "returned")
        }
    }

    fun summary(now: Long): MemberDigitalSummary {
        val carriage = approval.carriage
        if (now !in 0 until approval.expiresAt || approval.mandate.lapsesAt?.let { now >= it } == true ||
            approval.candidates.isEmpty() || approval.candidates.map { it.id }.distinct().size != approval.candidates.size ||
            carriage == null || carriage !in 0..Canonical.MAXIMUM_INTEGER) throw MemberFailure.Malformed
        var goods = 0L
        approval.candidates.forEach { candidate ->
            val choice = choice(candidate.id)
            if (candidate.valence != "offered" || choice == MemberDigitalChoice.UNDECIDED || candidate.quantity !in 1..Canonical.MAXIMUM_INTEGER || candidate.unitPrice !in 0..Canonical.MAXIMUM_INTEGER) throw MemberFailure.Malformed
            if (choice == MemberDigitalChoice.KEEP && candidate.givenBy == null) {
                val amount = try { Math.multiplyExact(candidate.quantity, candidate.unitPrice) } catch (_: ArithmeticException) { throw MemberFailure.Malformed }
                if (amount > Canonical.MAXIMUM_INTEGER || goods > Canonical.MAXIMUM_INTEGER - amount) throw MemberFailure.Malformed
                goods += amount
            }
        }
        if (goods > Canonical.MAXIMUM_INTEGER - carriage) throw MemberFailure.Malformed
        return MemberDigitalSummary(goods, carriage, goods + carriage)
    }
}

data class PreparedMemberDecision(
    val environment: MemberEnvironment,
    val session: MemberSessionInfo,
    val detail: MemberOfferDetail,
    val approval: MemberApproval,
    val decisions: List<Decision>,
    val canonical: String,
    val summary: MemberDigitalSummary,
) {
    companion object {
        fun create(environment: MemberEnvironment, session: MemberSessionInfo, detail: MemberOfferDetail, draft: MemberDigitalDraft, now: Long): PreparedMemberDecision {
            if (session.expiresAt <= now || detail.expiresAt <= now || detail.state != "presented" || detail.binding != "digital" ||
                session.household != detail.household || detail.presenter !in session.presenters || draft.approval.offer != detail.id || draft.approval.presenter != detail.presenter) throw MemberFailure.ScopeMismatch
            val decisions = draft.decisions(now)
            return PreparedMemberDecision(environment, session, detail, draft.approval, decisions, Canonical.decisions(detail.id, decisions), draft.summary(now))
        }
    }
}
